README.md
![](https://bitbucket.org/berkeleylab/checkv/raw/6d4448f738ac8549551c8ef9511afb05bc394813/logo.png)

[![PyPI](https://img.shields.io/pypi/v/checkv.svg?label=PyPI&color=green)](https://pypi.python.org/pypi/checkv)
[![Conda](https://img.shields.io/conda/vn/bioconda/checkv.svg?label=Conda&color=green)](https://anaconda.org/bioconda/checkv)
[![PyPI downloads](https://img.shields.io/pypi/dm/checkv?label=PyPI%20downloads&color=blue)](https://pypi.python.org/pypi/checkv)
[![Conda downloads](https://img.shields.io/conda/dn/bioconda/checkv.svg?label=Conda%20downloads&color=blue)](https://anaconda.org/bioconda/checkv)
[![Docker pulls](https://img.shields.io/docker/pulls/antoniopcamargo/checkv.svg?label=Docker%20pulls&color=blue)](https://hub.docker.com/r/antoniopcamargo/checkv)

CheckV is a fully automated command-line pipeline for assessing the quality of single-contig viral genomes, including identification of host contamination for integrated proviruses, estimating completeness for genome fragments, and identification of closed genomes.

The pipeline can be broken down into 4 main steps:

![](https://bitbucket.org/berkeleylab/checkv/raw/657fde9b1c696185a399456fbcbb4ca82066abb6/pipeline.png)

**A: Remove host contamination on proviruses**

* Genes are first annotated as viral or microbial based on comparison to a custom database of HMMs
* CheckV scans over the contig (5' to 3') comparing gene annotations and GC content between a pair of adjacent gene windows
* This information is used to compute a score at each intergenic position and identify host-virus breakpoints
* Host-virus breakpoints are identified with:
	* High scores (>1.2)
	* A minimum of 2 host-specific genes in the putative host region (for contigs with >=10 genes)
	* A minimum of 2 virus-specific genes in the putative viral region (for contigs with >=10 genes)
	* A minimum of 30% genes annotated as microbial in the putative host region

**B: Estimate genome completeness (2 algorithms)**

* AAI-based approach (accurate point estimate for genome completeness; pictured above)
	* First, proteins are compared to the CheckV genome database using AAI (average amino acid identity)
	* After identifying the top hits, completeness is computed as a ratio between the contig length (or viral region length for proviruses) and the length of matched reference genomes
	* A confidence level is reported based on the strength of the alignment and the length of the contig
	* Generally, high- and medium-confidence estimates are quite accurate and can be trusted
* HMM-based approach (estimated range for genome completeness)
	* Highly novel viruses may not match a CheckV genome with sufficient AAI (i.e. low-confidence estimate)
	* In these cases CheckV identifies the viral HMMs on the contig (see panel A) and compares the contig length with reference genomes sharing the same HMMs
	* CheckV then returns the estimated range for genome completeness (e.g. 35% to 60% completeness), which represents the 90% confidence interval based on the distribution of lengths of reference genomes with the same viral HMMs

**C: Predict closed genomes (3 signatures)**

* Direct terminal repeats (DTRs)
	* Repeated sequence of >20-bp at start/end of contig
	* Most trusted signature in our experience
	* May indicate circular genome or linear genome replicated from a circular template (i.e. concatamer)
* Proviruses
	* Viral region with predicted host boundaries at 5' and 3' ends (see panel A)
	* Note: CheckV will not detect proviruses if host regions have already been removed (e.g. using VIBRANT or VirSorter)
* Inverted terminal repeats (ITRs)
 	* Repeated sequence of >20-bp at start/end of contig (3' repeat is inverted)
	* Least trusted signature in our experience

CheckV will also report a confidence level based on comparison to completeness estimates (panel B):

* High-confidence: >90% estimated completeness
* Medium-confidence: 80-90% estimated completeness
* Low-confidence: <80% estimated completeness

For DTRs and ITRs, CheckV performs some additional filtering/checks:

* Ambiguous bases in repeat (e.g. "NNNNN"): <= 20% of repeat sequence with Ns
* Mode base frequency in repeat (e.g. "AAAAA"): <= 75% of repeat sequence composed of single base
* Maximum occurences of repeat sequence: <= 8 times per contig (removes highly repetetive sequences)
* Maximum kmer-frequency: <= 1.5 (removes contigs with the same genome repeated back-to-back)


**D: Summarize quality.**

Based on the results of A-C, CheckV generates a report file and assigns query contigs to one of five quality tiers (consistent with and expand upon the MIUViG quality tiers):

* Complete (high- or medium-confidence predictions; see panel C)
* High-quality (>90% completeness)
* Medium-quality (50-90% completeness)
* Low-quality (<50% completeness)
* Undetermined quality


## Installation

There are three methods to install CheckV:

- Using `conda`:

```bash
conda install -c conda-forge -c bioconda checkv
```

- Using `docker`: see section below

- Using `pip`:

```bash
pip install checkv
```

If you decide to install CheckV via `pip`, make sure you also have the following external dependencies installed:

- DIAMOND (v2.0.4)
- HMMER (v3.3)
- Prodigal (v2.6.3)

The versions listed above were the ones that were properly tested. Different versions may also work.

### CheckV database

If you install using `conda` or `pip` you will need to download the database:

```bash
checkv download_database ./
```

Or (to download and extract manually):

```bash
wget https://portal.nersc.gov/CheckV/checkv-db-v1.0.tar.gz
tar -zxvf checkv-db-v1.0.tar.gz
```

And update your environment (optional):

```bash
export CHECKVDB=/path/to/checkv-db-v1.0
```

Some users may wish to update the database using their own complete genomes:
```bash
checkv update_database /path/to/checkv-db /path/to/updated-checkv-db genomes.fna
```

## Quick start

There are two ways to run CheckV:

- Using individual commands for each step in the pipeline in the following order:

```bash
checkv contamination input_file.fna output_directory -t 16
checkv completeness input_file.fna output_directory -t 16
checkv complete_genomes input_file.fna output_directory
checkv quality_summary input_file.fna output_directory
```

- Using a single command to run the full pipeline:

```bash
checkv end_to_end input_file.fna output_directory -t 16
```

- For a full listing of checkv programs and options, use: `checkv -h` and `checkv <program> -h`

## Docker usage

- Pull the image (including the database)

```bash
docker pull antoniopcamargo/checkv
```

- Run the image

```bash
docker run -ti --rm -v "$(pwd):/app" antoniopcamargo/checkv end_to_end input_file.fna output_directory -t 16
```


## Automated testing

CheckV uses the [pytest](https://docs.pytest.org/en/latest/) framework for checking the accuracy of the outputs of its main functions. The testing pipeline executes the `contamination`, `completeness`, `complete_genomes` and `quality_summary` modules sequentially and then compares the output files with the ground truth.

To avoid introducing unwanted breaking changes, we recommend developers and users to execute the tests before committing any changes or submitting new PRs to the repository.

```bash
pytest --database=/path/to/checkv-db --threads=16
```

## Output files

#### quality_summary.tsv

This contains integrated results from the three main modules and should be the main output referred to. Below is an example to demonstrate the type of results you can expect in your data:


| contig_id | contig\_length | 	provirus | 	proviral\_length | 	gene_count | 	viral_genes | 	host_genes | 	checkv_quality | 	miuvig_quality | 	completeness | 	completeness\_method | 	complete\_genome_type | 	contamination | 	kmer_freq | 	warnings |
|---------|--------|--------|--------|--------|--------|--------|--------|--------|--------|--------|--------|--------|--------|--------|
| 1 | 	5325 | 	No | 	NA | 	11 | 	0 | 	2 | 	Not-determined | 	Genome-fragment | 	NA | 	NA | 	NA | 	0 | 	1 | 	no viral genes detected |
| 2 | 	41803 | 	No | 	NA | 	72 | 	27 | 	1 | 	Low-quality | 	Genome-fragment | 	21.99 | 	AAI-based (medium-confidence) | 	NA | 	0 | 	1 | flagged DTR	 |
| 3 | 	38254 | 	Yes | 	36072 | 	54 | 	23 | 	2 | 	Medium-quality | 	Genome-fragment | 	80.3 | 	HMM-based (lower-bound) | 	NA | 	5.7 | 	1 | 	 |
| 4 | 	67622 | 	No | 	NA | 	143 | 	25 | 	0 | 	High-quality | 	High-quality | 	100 | 	AAI-based (high-confidence) | 	NA | 	0 | 	1.76 | 	high kmer_freq |
| 5 | 	98051 | 	No | 	NA | 	158 | 	27 | 	1 | 	Complete | 	High-quality | 	100 | 	AAI-based (high-confidence) | 	DTR | 	0 | 	1 | 	 |

In the example, above there are results for 6 viral contigs:

* The first 5325 bp contig has no completeness prediction, which is indicated by 'Not-determined' for the 'checkv_quality' field. This contig also has no viral genes identified, so there's a chance it may not even be a virus.
* The second 41803 bp contig is classified as 'Low-quality' since its completeness is <50%. This is estimate is based on the 'AAI' method. Note that only either high- or medium-confidence estimates are reported in the quality_summary.tsv file. You can see 'completeness.tsv' for more details. This contig had a DTR, but it was flagged for some reason (see complete_genomes.tsv for details)
* The third contig is considered 'Medium-quality' since its completeness is estimated to be 80%, which is based on the 'HMM' method. This means that it was too novel to estimate completeness based on AAI, but shared an HMM with CheckV reference genomes. Note that this value represents a lower bound (meaning the true completeness may be higher but not lower than this value). Note that this contig is also classified as a provirus.
* The fourth contig is classified as High-quality based on a completness of >90%. However, note that value of 'kmer_freq' is 1.7. This indicates that the viral genome is represented multiple times in the contig. These cases are quite rare, but something to watch out for.
* The fifth contig is classified as Complete based on the presence of a direct terminal repeat (DTR) and has 100% completeness based on the AAI method. This sequence can condifently treated as a complete genome.


#### completeness.tsv

A detailed overview of how completeness was estimated:


| contig_id  | contig_length  | proviral_length  | aai\_expected_length  | aai_completeness  | aai_confidence  | aai_error  | aai\_num_hits  | aai\_top_hit  | aai_id  | aai_af  | hmm\_completeness_lower  | hmm\_completeness_upper  | hmm_hits  |
|---------|---------|---------|---------|---------|---------|---------|---------|---------|---------|---------|---------|---------|---------|
1  | 9837  | 5713  | 53242.8  | 10.7  | high  | 3.7  | 10  | DTR_517157  | 78.5  | 34.6  | 5  | 15  | 4  |
2  | 39498  | NA  | 37309  | 100.0  | medium  | 7.7  | 11  | DTR_357456  | 45.18  | 30.46  | 75  | 100 | 22 |
3  | 29224  | NA  | 44960.1  | 65.8  | low  | 15.2  | 17  | DTR_091230  | 39.74  | 19.54  | 52  | 70  | 10  |
4  | 23404  | NA  | NA  | NA  | NA  | NA  | 0  | NA  | NA  | NA  | NA  | NA  | 0 |

In the example, above there are results for 4 viral contigs:

* The first proviral contig has an estimated completeness of 10.7% using on the AAI-based method (100 x 5713 / 53242.8). The confidence for this estimate is high, based on a relative estimated error of 3.7%, which is in turn based on the aai\_id (average amino acid identity) and aai\_af (alignment fraction of contig) to the CheckV reference DTR_517517
* The second contig has a completeness of 100% using the AAI-based method and a completeness range of 75 - 100% using the HMM-based method. Note that the contig length is a bit longer than the expected genome length of 37,309 bp.
* The third contig is estimated to be 65.8% complete based on the AAI approach. However we can't trust this all that much since the aai_confidence is low (meaning the top hit based on AAI was fairly weak). To be conservative, we may wish to report the range of completeness (52-70%) based on the HMM approach
* The last contig doesn't have any hits based on AAI and doesn't have any viral HMMs, so there's nothing we can say about this sequence

#### contamination.tsv

A detailed overview of how contamination was estimated:


| contig_id | 	contig_length | 	total_genes | 	viral_genes | 	host_genes | 	provirus | 	proviral_length | 	host_length | 	region_types | 	region_lengths | 	region\_coords_bp | 	region\_coords_genes | 	region\_viral_genes | 	region\_host_genes |
|---------|---------|---------|---------|---------|---------|---------|---------|---------|---------|---------|---------|---------|---------|
| 1 | 	98051 | 	158 | 	27 | 	1 | 	No | 	NA | 	NA | 	NA | 	NA | 	NA | 	NA | 	NA | 	NA |
| 2 | 	38254 | 	54 | 	23 | 	2 | 	Yes | 	36072 | 	2182 | 	host,viral | 	1-2182,2183-38254 | 	1-2182,2183-38254 | 	1-4,5-54 | 	0,23 | 	2,0 |
| 3 | 	6930 | 	9 | 	1 | 	2 | 	Yes | 	3023 | 	3907 | 	viral,host | 	3023,3907 | 	1-3023,3024-6930 | 	1-5,6-9 | 	1,0 | 	0,2 |
| 4 | 	101630 | 	103 | 	7 | 	24 | 	Yes | 	28170 | 	73460 | 	host,viral,host | 	46804,28170,26656 | 	1-46804,46805-74974,74975-101630 | 	1-43,44-85,86-103 | 	0,7,0 | 	13,0,11 |

In the example, above there are results for 4 viral contigs:

* The first contig is not a predicted provirus
* The second contig has a predicted host region covering 2182 bp
* The third 6930 bp contig has a host region identified on the left side
* The fourth 101630 bp contig has 103 genes, including 7 viral and 24 host genes. CheckV identified two host-virus boundaries

#### complete_genomes.tsv

A detailed overview of putative complete genomes identified:


| contig_id | 	contig_length | 	prediction_type | 	confidence_level | 	confidence_reason | 	repeat_length | 	repeat_count |
|---------|---------|---------|---------|---------|---------|---------|
| 1 | 	44824 | 	DTR | 	high | 	AAI-based completeness > 90% | 	253 | 	2 |

RELEASE_NOTES.txt
Version 0.8.1
* Check for DIAMOND errors and DIAMOND output file (issues #15 and #39)

Version 0.8.0
* Check if input FASTA file is empty, wrongly formatted, or duplicated sequence IDs (issue #14, pull #5)
* Check whether the execution of hmmsearch was completed before skipping it (pull #5)
* Speed up kmer based genome copy number estimation (issue #26, pull #4)
* Add a --remove_tmp flag to quality_summary and end_to end modules (pull #3)

Version 0.7.0
* update to contamination module to reduce false positive proviruses
 -require at least 2 host genes in proviral host regions, except for contigs with <= 10 genes. reduces false positives in real data.
 -improved sliding window algorithm: trailing window grows from last breakpoint while leading window has fixed size
 -size of leading window is dynamically set: 30% of genes with min of 15 and max of 50
 -additional statistics reported for each host/viral region
* update to hmm-based completeness for more robust estimates
 -improved identification of 5% and 95% quantiles
 -use all HMMs to calculate completeness: each yields the 5%-95% completeness range, and these ranges are averaged after weighting each by the HMMs coefficient of variation (based on how variable reference genome sizes are for the HMM)
 -improves concordance with AAI-based completeness
* update to complete_genomes module (formerly called repeats)
 -added confidence levels, based on comparison to completeness and detection of artifacts

Version 0.6.0
* Bug fix in completeness module that suppressed HMM-based results
* HMM-based completeness now reported as range: 90% confidence interval = X-Y% complete
* Warnings added to quality_summary.tsv for range of conditions
* Minor parameter change for genome copy number estimation
* Terminal repeat sequence included in output file (e.g. GCACTTC...)
* Added terminal repeat metadata: ambig_freq and max_base_freq
* Dropped automated filtering of terminal repeats: up to user to decide how to handle flagged seqs
* Quality summary reports multiple types of termini when present (e.g. DTR & ITR on same contig)
* cleaned_contigs.fna split into viruses.fna and proviruses.fna
* Renamed field 'prophage' to 'provirus' in quality_summary.tsv

Version 0.5.1
* Added download_database module to automatically download latest version of db
* Added update_database module to update db with new genomes
* Added end_to_end module to run all CheckV modules
* Added documentation for output files in README
* Help text reformatted

Version 0.4.0
* New HMM-based completeness estimation
* New estimation of genome copy #
* Completeness capped at 100% in quality_summary.tsv output
* Dropped 'comments' field in quality_summary.tsv output
* Computation of aligned fraction fixed
* terminal_repeats module renamed to repeats (to accomonate repeat based estimation of genome copy #)
* All TRs now reported in repeats.tsv
* New field 'repeat_flagged' which indicates low complexity/repetetive repeats
* Added RELEASE_NOTES.txt file
* CheckV print program and database versions
* CheckV checks for input files in quality_summary module

Version 0.3.0
* Added documentation to README.md
* Support for compressed input files (gz, bz2, xz)
* Param change for contamination module (win size: 35 to 40, min % host genes 20% to 30%)
* Generate cleaned_contigs.fna


checkv/modules/completeness.py
import argparse
import bisect
import csv
import multiprocessing as mp
import operator
import os
import shutil
import time

import checkv
import numpy as np
from checkv import utility


class Genome:
    def __init__(self):
        pass


class Gene:
    def __init__(self):
        pass


def fetch_arguments(parser):
    parser.set_defaults(func=main)
    parser.set_defaults(program="completeness")
    parser.add_argument(
        "input",
        type=str,
        help="Input nucleotide sequences in FASTA format (.gz, .bz2 and .xz files are supported)",
    )
    parser.add_argument("output", type=str, help="Output directory")
    parser.add_argument(
        "-d",
        dest="db",
        type=str,
        required=False,
        metavar="PATH",
        help="Reference database path. By default the CHECKVDB environment variable is used",
    )
    parser.add_argument(
        "-t",
        dest="threads",
        type=int,
        default=1,
        metavar="INT",
        help="Number of threads to use for Prodigal and DIAMOND",
    )
    parser.add_argument(
        "--restart",
        action="store_true",
        default=False,
        help="Overwrite existing intermediate files. By default CheckV continues where program left off",
    )
    parser.add_argument(
        "--percent_of_top_hit", type=float, default=50.0, help=argparse.SUPPRESS,
    )
    parser.add_argument(
        "--max_aai", type=float, default=None, help=argparse.SUPPRESS,
    )
    parser.add_argument(
        "--exclude_identical",
        action="store_true",
        default=False,
        help=argparse.SUPPRESS,
    )
    parser.add_argument(
        "--exclude_list", type=str, default=None, metavar="PATH", help=argparse.SUPPRESS
    )
    parser.add_argument(
        "--quiet", action="store_true", default=False, help="Suppress logging messages",
    )


def set_defaults(args):
    key_values = [
        ("percent_of_top_hit", 50),
        ("max_aai", None),
        ("exclude_identical", False),
        ("exclude_list", None),
    ]
    for key, value in key_values:
        if key not in args:
            args[key] = value


def yield_query_alns(path):
    """
    Yields list of formatted blastp alignments per 'qname'
    """
    try:
        handle = utility.parse_blastp(path)
        alns = [next(handle)]
        query = alns[0]["qname"].rsplit("_", 1)[0]
        for r in handle:
            if r["qname"].rsplit("_", 1)[0] != query:
                yield query, alns
                alns = []
                query = r["qname"].rsplit("_", 1)[0]
            alns.append(r)
        yield query, alns
    except:
        return


def init_database(args):
    """
    Initialize genes, genomes, refs, and exclude data objects
    genes: dictionary of gene_id to gene object (from input)
    genomes: dictionary of genome_id to genome object (from input)
    refs: dictionary of checkv_id to genome object (from database)
    exclude: set of user-defined checkv_ids to exclude from AAI-based estimation
    """

    # read fna input
    genomes = {}
    for header, seq in utility.read_fasta(args["input"]):
        genome = Genome()
        genome.id = header.split()[0]
        genome.seq = seq
        genome.length = len(seq)
        genome.genes = 0
        genome.protlen = 0
        genome.viral_length = None
        genome.aai = []
        genome.hmms = set([])
        genome.hmm_completeness = (None, None)
        genome.expected_length = None
        genome.aai_completeness = None
        genome.aai_confidence = None
        genome.aai_error = None
        genomes[genome.id] = genome

    # read input genes
    genes = {}
    for header, seq in utility.read_fasta(args["faa"]):
        gene = Gene()
        gene.id = header.split()[0]
        gene.length = len(seq)
        gene.genome_id = gene.id.rsplit("_", 1)[0]
        genes[gene.id] = gene
        genomes[gene.genome_id].genes += 1
        genomes[gene.genome_id].protlen += gene.length

    # read prophage predictions, if exists
    p = os.path.join(args["output"], "contamination.tsv")
    if os.path.exists(p):
        for r in csv.DictReader(open(p), delimiter="\t"):
            if "genome_id" in r:  ## temporary fix due to changing field format
                r["contig_id"] = r["genome_id"]
            if "viral" in r["region_types"]:
                genomes[r["contig_id"]].viral_length = int(r["proviral_length"])
            else:
                genomes[r["contig_id"]].viral_length = None

    # read reference genomes
    refs = {}
    p = os.path.join(args["db"], "genome_db/checkv_reps.tsv")
    for r in csv.DictReader(open(p), delimiter="\t"):
        genome = Genome()
        genome.id = r["checkv_id"]
        genome.length = int(r["length"])
        genome.type = r["type"]
        genome.weight = (
            1 if r["type"] == "circular" else 2 if r["type"] == "genbank" else None
        )
        refs[genome.id] = genome

    # exclude references
    exclude = set([])
    if args["exclude_list"]:
        for l in open(args["exclude_list"]):
            exclude.add(l.rstrip())

    # read hmm info
    hmms = {}
    p = os.path.join(args["db"], "hmm_db/genome_lengths.tsv")
    for r in csv.DictReader(open(p), delimiter="\t"):
        r["genomes"] = int(r["genomes"])
        r["cv"] = float(r["cv"])
        r["lengths"] = [int(_) for _ in r["lengths"].split(",")]
        hmms[r["hmm"]] = r

    return genes, genomes, refs, exclude, hmms


def compute_aai(blastp_path, out_path, genomes, genes, refs):
    """ Compute AAI to reference genomes
    1. loop over alignment blocks from <blastp_path> (1 per query)
    2. compute average aa identity to each reference genome
    3. write all hits to disk
    """

    with open(out_path, "w") as out:
        header = [
            "query",
            "target",
            "aligned_genes",
            "percent_genes",
            "aligned_length",
            "percent_length",
            "identity",
            "score",
        ]
        out.write("\t".join(header) + "\n")

        for query, alns in yield_query_alns(blastp_path):

            # get best hits for each gene to each target genome
            hits = {}
            for r in alns:
                target = r["tname"].rsplit("_", 1)[0]
                # store gene aln
                if target not in hits:
                    hits[target] = {}
                if r["qname"] not in hits[target]:
                    hits[target][r["qname"]] = r
                elif r["score"] > hits[target][r["qname"]]["score"]:
                    hits[target][r["qname"]] = r

            # compute aai
            aai = []
            for target, alns in hits.items():
                pids = [r["pid"] for r in alns.values()]
                lens = [genes[_].length for _ in alns]
                aligned_length = sum(lens)
                aligned_genes = len(alns)
                identity = round(
                    sum(x * y for x, y in zip(pids, lens)) / aligned_length, 2
                )
                percent_length = round(
                    100.0 * aligned_length / genomes[query].protlen, 2
                )
                percent_genes = round(100.0 * aligned_genes / genomes[query].genes, 2)
                score = round(identity * aligned_length / 100, 2)
                row = [
                    query,
                    target,
                    aligned_genes,
                    percent_genes,
                    aligned_length,
                    percent_length,
                    identity,
                    score,
                ]
                aai.append(row)
            if not aai:
                continue

            # write all hits to disk
            aai = sorted(aai, key=operator.itemgetter(-1), reverse=True)
            top_score = max(_[-1] for _ in aai)
            for row in aai:
                score = row[-1]
                out.write("\t".join(str(_) for _ in row) + "\n")


def store_aai(args, genomes, exclude, refs):
    """
    1. Loop over AAI records from disk
    2. Format numeric values
    3. Filter records based on exclusion criterea
    4. For each query, store all AAI records with alignment scores within 50% of top hit
    """

    for r in csv.DictReader(open(args["aai"]), delimiter="\t"):

        r["identity"] = float(r["identity"])
        r["aligned_genes"] = float(r["aligned_genes"])
        r["aligned_length"] = int(r["aligned_length"])
        r["score"] = float(r["score"])
        r["percent_length"] = float(r["percent_length"])

        if args["max_aai"] is not None and r["identity"] > args["max_aai"]:
            continue
        elif r["target"] in exclude:
            continue
        elif r["target"] not in refs:
            continue
        elif (
            args["exclude_identical"]
            and r["identity"] == 100
            and r["percent_length"] == 100
        ):
            continue

        if len(genomes[r["query"]].aai) == 0:
            genomes[r["query"]].aai.append(r)
        else:
            top_score = genomes[r["query"]].aai[0]["score"]
            if (
                100.0 * (top_score - r["score"]) / top_score
                <= args["percent_of_top_hit"]
            ):
                genomes[r["query"]].aai.append(r)


def take_closest(myList, myNumber):
    """
    Assumes myList is sorted. Returns closest value to myNumber.
    If two numbers are equally close, return the smallest number.
    """

    myNumber = float(myNumber)
    pos = bisect.bisect_left(myList, float(myNumber))
    if pos == 0:
        return myList[0]
    if pos == len(myList):
        return myList[-1]
    before = myList[pos - 1]
    after = myList[pos]
    if (after - myNumber) < (myNumber - before):
        return after
    else:
        return before


def lookup_error_rate(genome, error_rates, error_keys):
    """
    Returns estimated error rate for genome based on AAI, AF, and length
    """

    key1 = take_closest(error_keys["length"], genome.length)
    key2 = take_closest(error_keys["aai"], genome.aai[0]["identity"])
    key3 = take_closest(error_keys["cov"], genome.aai[0]["percent_length"])
    error = error_rates[key1, key2, key3]
    confidence = (
        "low"
        if error == "NA"
        else "high"
        if error <= 5
        else "medium"
        if error <= 10
        else "low"
    )
    return error, confidence


def aai_based_completeness(args, genomes, exclude, refs):
    """
    0. Store error rates
    1. Fetch top database hit
    2. Get estimation error based on AAI and AF to top hit as well as contig length
    3. Compute expected genome length as weighted average of all hits with an alignment score within 50% of top hit
    4. Compute completeness as ratio of contig length (or viral length for proviruses) to expected length
    """

    # store aai
    store_aai(args, genomes, exclude, refs)

    # store error rates
    error_rates = {}
    p = os.path.join(args["db"], "genome_db/checkv_error.tsv")
    for r in csv.DictReader(open(p), delimiter="\t"):
        key = int(r["length"]), int(r["aai"]), int(r["cov"])
        error_rates[key] = float(r["error"]) if int(r["count"]) > 100 else "NA"

    # list keys for lookup
    error_keys = {
        "length": sorted(list({_[0] for _ in error_rates})),
        "aai": sorted(list({_[1] for _ in error_rates})),
        "cov": sorted(list({_[2] for _ in error_rates})),
    }

    # compute completeness
    for genome in genomes.values():
        if len(genome.aai) > 0:
            est_error, confidence = lookup_error_rate(genome, error_rates, error_keys)
            lengths = [refs[_["target"]].length for _ in genome.aai]
            weights = [refs[_["target"]].weight * _["score"] for _ in genome.aai]
            genome.expected_length = sum(l * w for l, w in zip(lengths, weights)) / sum(
                weights
            )
            query_length = (
                genome.viral_length
                if genome.viral_length is not None
                else genome.length
            )
            genome.aai_completeness = min(
                [100.0 * query_length / genome.expected_length, 100.0]
            )
            genome.aai_confidence = confidence
            genome.aai_error = est_error


def hmm_based_completeness(args, genomes, hmms, annotation_path):
    """
    1. Store HMM information (name, genome lengths, CV)
    2. List HMMs hitting each genome
    3. Identify the HMM whose database targets display the least genome size variation (i.e. smallest CV)
    4. Compare the contig length (or viral region on proviruses) to the database target lengths
    5. Estimate the 90% confidence interval of completeness
    """

    # list hmms hitting each genome
    for r in csv.DictReader(open(annotation_path), delimiter="\t"):
        if r["hmm_name"] in hmms and hmms[r["hmm_name"]]["genomes"] >= 10:
            genomes[r["contig_id"]].hmms.add(r["hmm_name"])

    # estimate minimum completeness
    for genome in genomes.values():

        # check at least 1 hmm
        if len(genome.hmms) == 0:
            continue

        # fetch genome length (viral region only)
        query_length = (
            genome.viral_length if genome.viral_length is not None else genome.length
        )

        # get weighted completeness range
        comps = []
        for hmm in genome.hmms:
            len_q1 = np.quantile(hmms[hmm]["lengths"], 0.05)
            len_q2 = np.quantile(hmms[hmm]["lengths"], 0.95)
            comp_q1 = round(min([100 * query_length / len_q2, 100.0]), 2)
            comp_q2 = round(min([100 * query_length / len_q1, 100.0]), 2)
            cv = hmms[hmm]["cv"]
            weight = min([1 / cv, 50]) if cv != 0 else 50
            comps.append([weight, comp_q1, comp_q2])
        weight_total = sum(weight for weight, comp_q1, comp_q2 in comps)
        comp_lower = (
            sum(weight * comp_q1 for weight, comp_q1, comp_q2 in comps)
            / weight_total
        )

        comp_upper = (
            sum(weight * comp_q2 for weight, comp_q1, comp_q2 in comps)
            / weight_total
        )

        genome.hmm_completeness = (comp_lower, comp_upper)


def main(args):

    program_start = time.time()
    set_defaults(args)
    logger = utility.get_logger(args["quiet"])
    utility.check_executables(["prodigal", "diamond"])
    args["db"] = utility.check_database(args["db"])
    args["tmp"] = os.path.join(args["output"], "tmp")

    if args["restart"] and os.path.exists(args["tmp"]):
        shutil.rmtree(args["tmp"])
    if not os.path.exists(args["output"]):
        os.makedirs(args["output"])
    if not os.path.exists(args["tmp"]):
        os.makedirs(args["tmp"])

    utility.check_fasta(args["input"], args["tmp"])

    logger.info(f"\nCheckV v{checkv.__version__}: completeness")

    args["faa"] = os.path.join(args["tmp"], "proteins.faa")
    if os.path.exists(args["faa"]):
        logger.info("[1/8] Skipping gene calling...")
    else:
        logger.info("[1/8] Calling genes with Prodigal...")
        utility.call_genes(args["input"], args["output"], args["threads"])

    logger.info("[2/8] Initializing queries and database...")
    genes, genomes, refs, exclude, hmms = init_database(args)

    args["blastp"] = os.path.join(args["tmp"], "diamond.tsv")
    args["diamond_checkpoint"] = os.path.join(args["tmp"], "diamond_checkpoint")
    if os.path.exists(args["diamond_checkpoint"]):
        logger.info("[3/8] Skipping DIAMOND blastp search...")
    else:
        logger.info("[3/8] Running DIAMOND blastp search...")
        utility.run_diamond(args["blastp"], args["db"], args["faa"], args["threads"])
        with open(args["diamond_checkpoint"], "w") as fout:
            pass

    args["aai"] = os.path.join(args["tmp"], "aai.tsv")
    if os.path.exists(args["aai"]):
        logger.info("[4/8] Skipping AAI computation...")
    else:
        logger.info("[4/8] Computing AAI...")
        compute_aai(args["blastp"], args["aai"], genomes, genes, refs)

    logger.info("[5/8] Running AAI based completeness estimation...")
    aai_based_completeness(args, genomes, exclude, refs)

    annotation_path = os.path.join(args["tmp"], "gene_features.tsv")
    if os.path.exists(annotation_path):
        logger.info("[6/8] Running HMM based completeness estimation...")
        hmm_based_completeness(args, genomes, hmms, annotation_path)
    else:
        logger.info(
            "[6/8] Skipping HMM based completeness estimation: requires output from 'checkv contamination'"
        )

    logger.info("[7/8] Determining genome copy number...")
    with mp.Pool(args["threads"]) as pool:
        kmer_freq_list = pool.map(utility.get_average_kmer_freq, genomes.values())
    for kmer_freq, genome in zip(kmer_freq_list, genomes.values()):
        genome.kmer_freq = kmer_freq

    logger.info("[8/8] Writing results...")
    p = os.path.join(args["output"], "completeness.tsv")
    fields = [
        "contig_id",
        "contig_length",
        "proviral_length",
        "aai_expected_length",
        "aai_completeness",
        "aai_confidence",
        "aai_error",
        "aai_num_hits",
        "aai_top_hit",
        "aai_id",
        "aai_af",
        "hmm_completeness_lower",
        "hmm_completeness_upper",
        "hmm_num_hits",
        "kmer_freq",
    ]
    with open(p, "w") as out:
        out.write("\t".join(fields) + "\n")
        for genome in genomes.values():
            rec = {
                "contig_id": genome.id,
                "contig_length": genome.length,
                "viral_length": genome.viral_length,
                "aai_expected_length": genome.expected_length,
                "aai_completeness": genome.aai_completeness,
                "aai_confidence": genome.aai_confidence,
                "aai_error": genome.aai_error,
                "aai_num_hits": len(genome.aai),
            }
            if len(genome.aai) > 0:
                rec["aai_top_hit"] = genome.aai[0]["target"]
                rec["aai_id"] = genome.aai[0]["identity"]
                rec["aai_af"] = genome.aai[0]["percent_length"]
            rec["hmm_completeness_lower"] = genome.hmm_completeness[0]
            rec["hmm_completeness_upper"] = genome.hmm_completeness[1]
            rec["hmm_num_hits"] = len(genome.hmms)
            rec["kmer_freq"] = genome.kmer_freq
            row = [
                str(rec[f]) if f in rec and rec[f] is not None else "NA" for f in fields
            ]
            out.write("\t".join(row) + "\n")

    # done!
    logger.info("Run time: %s seconds" % round(time.time() - program_start, 2))
    logger.info("Peak mem: %s GB" % round(utility.max_mem_usage(), 2))
